### PR TITLE
Add support for detection delay

### PR DESCRIPTION
**Pull request by francoisforster**
Initial implementation for https://github.com/elementary-data/elementary/issues/911

Since I'm not fully familiar with the elementary project as a developer, I'm expecting a lot of feedback and wanted to get this out before I spend too much time on the implementation.

One thing is that I added detection_delay_hours as a parameter as it didn't seem likely the delay would need specifying different time intervals.
I also implemented it as a shift in both the start and end time in the buckets. This matches my particular use-case but wanted to see if there are drawbacks compared to only changing the end time.

Also having issue getting tests to run successfully locally. Not sure why that is, but I'll keep trying to get them to succeed.

I'm sure I'm missing a bunch of things, like adding documentation, but that can be tackled as a follow-up.

### CODE REVIEW

**Review by elongl — COMMENTED** [review]
What an amazing implementation!
My main thought is that the interface should be similar to that of the [time bucket](https://docs.elementary-data.com/guides/anomaly-detection-configuration/time-bucket) such as:
```yaml
detection_delay:
  period: hour
  count: 4
```
What do you think?
Afterwards I think we'll head to testing as the code looks quite solid 👌🏽

**Comment by francoisforster** [comment]
@elongl I made the changes to match your request

File: integration_tests/tests/test_freshness_anomalies.py [hunk 1]
```diff
@@ -47,6 +49,7 @@ def _get_test_config(self, config: FreshnessAnomaliesConfig) -> dict:
             days_back=config.days_back,
             backfill_days=config.backfill_days,
             time_bucket=dict(period=config.period, count=1),
+            detection_delay_hours=dict(period='hour', count=config.detection_delay_hours)
```
Comment (ofek1weiss) [inline integration_tests/tests/test_freshness_anomalies.py:52]: shouldnt it be `detection_delay`?

Comment (francoisforster) [reply integration_tests/tests/test_freshness_anomalies.py:52]: yes, forgot to change it.

Comment (ofek1weiss) [reply integration_tests/tests/test_freshness_anomalies.py]: Great!

**Comment by ofek1weiss** [comment]
now it seems that all the tests in `test_freshness_anomalies` other than the delay one fail...
this is the exception they are getting:
```
detection_delay.count has to be larger than 0, got: 0
```
i think you can fix it by doing the following:
```
detection_delay=dict(...) if config.detection_delay_hours else None
```
you can try running the tests yourself (guild [here](config.detection_delay_hours)) if you would like :)

**Comment by francoisforster** [comment]
@ofek1weiss I can't seem to run the tests locally for some reason. Sorry about the churn. I've changed the config to allow 0 delay

**Comment by ofek1weiss** [comment]
@francoisforster seems like the test `TestFreshnessAnomalies::test_stop_with_delay[hourly]` fails for spark db,
do you have an idea as to why should it be the case?

**Comment by francoisforster** [comment]
Hmm, I don't know. Main branch tests fail locally for me against BQ and Postgres and I can't really test against spark db.

**Comment by francoisforster** [comment]
Nothing in my PR is datastore specific. All the logic is done in python

**Comment by ofek1weiss** [comment]
@francoisforster thats really bizarre.
If you can, id appreciate it if you try to run the tests on spark using [this guild](https://github.com/dbt-labs/dbt-spark#running-locally) to create a local spark instance (basically `git clone` and `docker compose up -d`) 🙏 .
I will look into it as well, but it might take some time

**Comment by ofek1weiss** [comment]
Hi @francoisforster, still looking into it.
in the meantime, can you resolve the conflicts with `master` (there are currently conflicts)? there are some changes there to the tests that might help us get to the bottom of the issue faster

**Comment by francoisforster** [comment]
I rebased.

Running the freshness tests for me from master branch (pytest tests/test_freshness_anomalies.py):
```
===================================================================================================== short test summary info ======================================================================================================
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_anomalyless_table[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_faster_rate[hourly] - AssertionError: assert 'fail' == 'pass'
============================================================================================= 2 failed, 14 passed in 321.84s (0:05:21) =============================================================================================
```

from my branch against Postgres:
```
===================================================================================================== short test summary info ======================================================================================================
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_anomalyless_table[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_stop_with_delay[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_faster_rate[hourly] - AssertionError: assert 'fail' == 'pass'
============================================================================================= 3 failed, 17 passed in 399.03s (0:06:39) =============================================================================================

```

against BigQuery
```
===================================================================================================== short test summary info ======================================================================================================
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_anomalyless_table[hourly] - AssertionError: assert 'error' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_stop_with_delay[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_faster_rate[hourly] - AssertionError: assert 'fail' == 'pass'
============================================================================================ 3 failed, 17 passed in 1225.18s (0:20:25) =============================================================================================

```

against Spark
```
===================================================================================================== short test summary info ======================================================================================================
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_anomalyless_table[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_stop_with_delay[hourly] - AssertionError: assert 'fail' == 'pass'
FAILED tests/test_freshness_anomalies.py::TestFreshnessAnomalies::test_faster_rate[hourly] - AssertionError: assert 'fail' == 'pass'
============================================================================================= 3 failed, 17 passed in 400.17s (0:06:40) =============================================================================================
```

There's obviously something going on with the hourly tests that's not due to my changes...

**Comment by ofek1weiss** [comment]
yeah it looks like it...
seems the hourly tests might be a bit too sensitive ATM, Im on it 👍

**Comment by francoisforster** [comment]
Thanks @dapollak . I'm away at the moment but I'll rebase as soon as i get back

**Comment by dapollak** [comment]
Hey again @francoisforster,
I've merged your contribution through [this](https://github.com/elementary-data/dbt-data-reliability/pull/565) branch.

Thanks !!

**Comment by francoisforster** [comment]
@dapollak Thanks for merging it!

**Pull request merged**